### DISCUSSION TITLE
React Streaming SSR/Server Components

### DISCUSSION THREAD

**Discussion by Ephem** [TanStack/query#4623] [Ideas]
# Improve React Query hydration, Streaming SSR & Server Components

> Update 2, TLDR: This post tries to outline a rough early draft API for integrating React Query with all the new React APIs while preserving a familiar API and trying to minimize the amount of new things RQ users needs to learn to leverage them, as well as outlining some todos to get there.

> Update: This a pretty complex and advanced subject where a lot of things are still changing/undecided and is as of yet undocumented and I didn't include a lot of "why's" or explanations in this post to keep it manageable. I tried to include resources that provides _some_ context, and please ask and comment on anything and I'll try to expand on it if I can!

For everything below, we also need to figure out a holistic API that makes sense, that we can build on in the future and with a solid migration path. Hopefully(/probably) APIs won't change too much/be incrementally adoptable, I _think_ most work happens in the background, but that there will be new patterns to learn for users that want to leverage new features.

This is not (yet) an API design document, but I've added some examples of my current thinking at the bottom.

Note that I think a big goal of Query is to have a framework agnostic core to be able to support other frameworks than React. It already has a great custom cache, so the plan is to keep building on that and not really rely on the upcoming React cache APIs except where we have to, for specific things like request scoped things, the fetch-deduplication etc. This might be subject to change of course.

> The [Suspense SSR Architecture post](https://github.com/reactwg/react-18/discussions/37) as well as [this post (esp the header "Hydration Data")](https://github.com/reactwg/react-18/discussions/114) in the React 18 WG has some great context to this subject, as well as [this comment](https://github.com/reactwg/react-18/discussions/121#discussioncomment-2207678).

> Note: If you have an itch to play around/experiment with RSCs and RQ early, there is a [WIP PR](https://github.com/TanStack/query/pull/4568) by @efilion that documents what is currently possible.

> Note: If you want to read some earlier thinking on the subject, [this discussion](https://github.com/TanStack/query/discussions/3173) has some history, but is no longer current/up to date.

## General stuff

* Add `'use client'` in the correct places (minor?)
* Refactor Suspense to `use()` instead of throwing Promises (minor/major?)
  * Can we/do we want to keep the old behaviour for older versions of React?
* `useSyncExternalStore` should have a correct/frozen `getServerSnapshot` (minor)
* Hydration implementation details (major)
  * Step 1: Should still happen in render for _new_ queries (and queries with no observers?), but be scheduled for a `useEffect` for queries that someone is actively listening to since this results in re-renders/observable side effects
  * Step 2 (optional): When hydrating in `useEffect` _also_ add the new data to a context for that subtree, refactor all reads to patch in data from that context. That subtree safely gets the new data during the same render instead of after, "true" cache and thus anything listening outside of that subtree gets updated via the `useEffect` (This might be overkill, we should evaluate if step 1 is good enough first)
* Suggestion: Rename `<Hydrate>` to `<HydrationBoundary>` (? open to other suggestions, and we can deprecate and keep the current one for migration purposes), for three reasons:
  * It's a descriptive rather than imperative name
  * If we implement Step 2 above, the name better explains why there are two different behaviours for in-tree and out-tree
  * If we can support `useQuery(..., { suspense: true })` with SSR Streaming (see below), we might need to support some kind of "automatic, in the background de/rehydration" tied to a specific position in the tree. Also, matching a `Suspense(Boundary)`, `ErrorBoundary` and `HydrationBoundary` together feels like it makes sense?
  * This said, we might want to hold off on this until we know more
* Figure out what to do with the imperative `hydrate()` (should stay for other frameworks) and the hook `useHydrate` (deprecate/remove?)

## Server Side Rendering

* Figure out/document if there is a way for `useQuery(..., { suspense: true })` to work out of the box as long as you buffer all the content instead of streaming? Pretty sure this is possible in custom setups, but not sure about metaframeworks.
* Figure out if there will be a way for `useQuery(..., { suspense: true })` to work with streaming SSR (without `await prefetch()` that same query before rendering). Missing primitives in React? (Linked question in the RSC WG: [#3](https://github.com/reactwg/server-components/discussions/3))
  * If this is not supported by React out of the box, is there a way to implement this at the library level as an out of band channel or something? (Probably not, at least without meta-framework support, maybe this is a (meta-)framework level concern to start with)

## Server Components

* Implement support for de/rehydrating Promises. This will mean rendering can start before all data has loaded (minor/major?)
  * Promise de/rehydration is not yet implemented in React, so we are blocked on this until then
  * This will likely be in the core, how does it affect other frameworks? Are there ways for other frameworks to leverage this that we should consider when designing it?
* Figure out what API we want in Server Components. Is a `queryClient` even necessary here, or should we implement a lightweight `prefetch()` API instead?
  * If we still want/need a `queryClient` to avoid duplication etc, we probably want to use the per request `cache()` API that is for now only documented in the [Next beta docs](https://beta.nextjs.org/docs/data-fetching/caching#per-request-caching). Not sure that this is necessary though, especially as a first step since it probably adds a bunch of complexity.

## API Examples

This is not thought through or fleshed out yet, just early spitballing to start a discussion. It all depends a lot on what final low level APIs will be available from React(/the metaframeworks) and what they will look like, but I consider it a decent guess based on what we know.

### Basic hydration

No Suspense, no streaming, no RSCs, just what's always been available. 

```jsx
import { prefetchQuery, HydrationBoundary } from '@tanstack/react-query';

// Will work both for initial SSR and for page navigations
// (page1->page2) where loaders are still executed serverside
// This could be a custom framework, Remix loaders,
// Next getServerSideProps or whatever
export async function loader() {
  // Note that we have no queryClient here, it's still supported, for
  // example for the client side case, but shouldn't be necessary here
  const dehydratedQuery = await prefetchQuery(...);

  // Could support prefetchQueries() too,
  // or HydrationBoundary could support either a single query,
  // or an array of queries, that way you could just do Promise.all()
  // and hydrate the result
  // Key here is that if you await all Promises here, no Promise
  // de/rehydration happens, the dehydrated state is pure JSON
  // We might want some way to guarantee this?
  
  // We are still expecting the framework to handle actual serialization here
  return { props: { dehydratedQuery }};
}

function PageComponent({ dehydratedQuery, children }) {
  return (
    <HydrationBoundary state={dehydratedQuery}>
      {children}
    </HydrationBoundary>
  )
}

// Deeper in the tree
function SomeComponent() {
  const { data } = useQuery(...);
  // or, but guaranteed to never suspend if await prefetchQuery
  // has always been called:
  const { data } = useQuery(..., { suspense: true });

  // ...
}
```

### Streaming SSR without prefetching

This probably needs primitives from React we don't have or know anything about yet, but those should(?) be transparent to the user, so we can still imagine an API.

```jsx
// Higher in the tree
function Page({ children }) {
  return (
    // Note: No state, in this case it just provides some
    // behind the scenes setup, might not be necessary
    <Suspense fallback="Loading...">
      <HydrationBoundary>{children}</HydrationBoundary>
    </Suspense>
  );
}

// Deeper in the tree
function SomeComponent() {
  // This will de/rehydrate the Promise/data in the background
  const { data } = useQuery(..., { suspense: true });

  // ...
}
```

Optionally, the `Page` component above could start a `prefetchQuery`, but not `await` it, just to kick off loading earlier. Some details to work out around that case (does it need to be `queryClient.prefetchQuery` in this case, or can we rely on the fetch cache here as well as in RSCs, should/can that prefetch happen in a `loader` function before rendering, or in render (since it's an idempotent side effect), etc?)

### Server Components

```jsx
export default async ServerComponent({children}) {
  // Note that in this case, we are not awaiting the
  // prefetch, so the thing that is inside the dehydratedQuery
  // is a Promise that is passed down to the "client", automatically
  // resolved when data comes in etc
  const dehydratedQuery = prefetchQuery(...);

  // Just do illustrate:
  const dehydratedQuery2 = prefetchQuery(...);

  return (
    <Suspense fallback="Loading...">
      <HydrationBoundary state={[dehydratedQuery, dehydratedQuery2]}>
        {children}
      </HydrationBoundary>
    </Suspense>
  );
}

// Deeper client component
function SomeComponent() {
  // Since we've hydrated an active Promise here, it will
  // suspend without a new clientside fetch and the Suspense
  // fallback will render. This will happen both during the
  // server rendering and in the browser.
  // When the Promise on the server resolves, React will take
  // care of resolving and rerendering correctly (hopefully, I think)
  const { data } = useQuery(..., { suspense: true });

  // ...
}
```

## What did I miss?

This is somewhat of a braindump on my current thinking, it's not well researched or thought through in all aspects, so take it as a starting point. Please challenge everything you can think of, tell me where I'm wrong, ask questions on anything that is unclear and join in the discussion!

**Comment by TkDodo** [comment]
the `Server Components` example looks pretty cool. If we could make it work that way, it would make passing props down to a client component unnecessary 👍

> Add 'use client' in the correct places (minor?)

yes this should be doable as a minor

> Refactor Suspense to use() instead of throwing Promises (minor/major?)
Can we/do we want to keep the old behaviour for older versions of React?

we can do this as soon as `use` becomes non-experimental I guess ? With v5, we will drop support for react < 18, so we should be good once `use` makes it to 18. This could pan out nicely with our v5 timeline.

> Suggestion: Rename <Hydrate> to <HydrationBoundary>

like it 👍 

> Figure out what to do with the imperative hydrate() (should stay for other frameworks) and the hook useHydrate (deprecate/remove?)

my first instinct is: `hydrate` should stay (`query-core`) because we also use it for other things like persisters. `useHydrate` can go.

---

> useQuery(..., { suspense: true })

one bigger question from my side would be: Do we want a dedicated `useSuspenseQuery` (and `useSuspenseInfiniteQuery` hook instead of the `suspense:true` boolean?

This would be beneficial for TypeScript users because right now, `data` is always `T | undefined` (and not `T`). We could change types for those hooks, and also e.g. disallow the `enabled` option to streamline the types with what people want mostly when using suspense: not to check for undefined data 😅

**Reply by Ephem** [reply]
> it would make passing props down to a client component unnecessary

Yeah, I guess another way of saying it would be that the `HydrationBoundary` _is_ the client component so to speak, and then that can render either more server components (works since they are passed in from above), or other client components depending on what is passed in as children.

This also reminds me, one thing I didn't cover in the post above is the case where you want to do something like this in a Server Component, and use the data immediately:

```jsx
async function ServerComponent() {
  const { data } = useQuery(...);
}
```

What should the API for that look like? If we support `useQuery` that relies on a `queryClient` currently which would be nice to avoid in RSCs if we can (at least to start with), but having another API could feel cumbersome. Something to figure out.

> With v5, we will drop support for react < 18, so we should be good once use makes it to 18. This could pan out nicely with our v5 timeline.

Lets hope! :+1: I guess this would mean we would require React 18.X, where X is the version where `use` was introduced though?

> `useSuspenseQuery`

Agreed. I was considering including that here, but it wasn't critical for any of the other things and I knew there were motivations I had forgotten about for changing the name. I like that this gives us the opportunity to tweak the API for things like `enabled`, we should think about if there are anything else we want to change at the same time I guess? The possible loading states obviously changes, but do we want to make any changes to error handling for example?

Should we remove the `suspense: true`-option in v5 in that case? How painful a migration would this be for people that have configured this at a global level for their apps? (Probably not very as long as we don't change too much, codemod perhaps?)

**Reply by Ephem** [reply]
About the case:

```jsx
async function ServerComponent() {
  const { data } = useQuery(...);
}
```

The problem here is that since this is a Server Component, it wouldn't rerender if we refetch fresher data on the client, any markup we rendered in the Server Component would still be stale. As a library, I'm not sure we _can_ revalidate Server Components as it stands now, since this is more a router level concern? If something in the key of that `useQuery` _also_ revalidated the Server Component though, it would probably work. A revalidates both B (SC) and C (the clientside query cache) so to speak.

But whatever we do at a library level here, `useQuery` in an SC would never mean/work the same as in a client component, so might make sense to design the API around that and have something else instead. Possibly just `const { data } = await prefetchQuery(...)` and use the data as is. I think this is the case I've thought least about so far.

**Reply by TkDodo** [reply]
> HydrationBoundary is the client component

yes, absolutely the right way to think about it 👍 

> where you want to do something like this in a Server Component

in a server component, you would just do:

```
const data = queryClient.fetchQuery({ queryKey, queryFn })
```

instead of `useQuery`

> I like that this gives us the opportunity to tweak the API for things like enabled, we should think about if there are anything else we want to change at the same time I guess? The possible loading states obviously changes, but do we want to make any changes to error handling for example?

It would default to using errorBoundaries, but you can still change that per query. I think that's fine.

> How painful a migration would this be for people that have configured this at a global level for their apps?

`suspense: true` was clearly marked as _experimental_, so I don't think the migration strategy should play too much of a role. What I like about the separation is that we can move a bunch of suspense only code to the separate function, which can be tree-shaked if you don't use suspense.

There is an edge case where if you cancel a query while it is fetching, it goes to loading state with undefined data. I don't know what would happen here with suspense though - it probably would just suspend again?

**Reply by Ephem** [reply]
> `const data = queryClient.fetchQuery({ queryKey, queryFn })`

Oh yeah, I should have written this instead of the hook form of course. :sweat_smile: Maybe I'm fighting too hard against having a `queryClient` in Server Components, I do have a few arguments, but I think mostly it's just a feeling that avoiding stateful abstractions shared between Server Components might be a good thing if we can, but I might very well be wrong. Definitely something to think more on and figure out.

**Reply by TkDodo** [reply]
I think you would just _create_ a `new QueryClient` where you need to - I also wouldn't share a client on the server

**Comment by amannn** [comment]
This is really interesting, thanks for sharing! I'd be curious if you have ideas how existing APIs that rely on React Context will work on the server side (e.g. `QueryClientProvider`)? [RSC don't support React Context](https://beta.nextjs.org/docs/rendering/server-and-client-components#context) at this point.

I just saw [the new SSR guide](https://github.com/TanStack/query/blob/9487691e37a7d04614c64a82bb3ac86a14ebb61d/docs/guides/ssr.md) where you're using the `cache()` API for this use case. So I guess you could use `getQueryClient` instead of `useQuery` in RSC's, but it's a bit unfortunate that you have to use different APIs depending on the environment, no?

**Reply by Ephem** [reply]
As I stated and also discussed with @TkDodo in other comments here, I'm not even sure we _need_ a `queryClient` in RSCs. A lot simplified, I see the `queryClient`s job as twofold:

* Sync state between different components
  * In RSCs, because `fetch` is deduplicated automatically, data will be in sync between RSCs anyway
* Handle subscriptions to data and trigger rerenders when data is refetched
  * Not applicable to RSCs

So we might be able to go for a lot simpler model here to avoid complexity.

Note though that Server Components aren't the only thing that runs on the server, the client components are still rendered on the server as well, and there the context and therefor `queryClient` will work as usual.

> it's a bit unfortunate that you have to use different APIs depending on the environment, no?

Depends, a common API for different environments is great if it matches the mental model of that environment, but RSCs might be such a different thing to think about that having a _slightly_ different API that matches it's uniqueness (that is, simpler, perhaps without a `queryClient` and geared towards prefetching) might not be a bad thing.

**Reply by Ephem** [reply]
Just wanted to expand on the `queryClient` bit and different mental models a bit more. We are very used to thinking in a top-down sense when it comes to React. We set up some Providers at the top and trust they are always there, providing their context. With Server Components, that might not always be the case (in the future?). What if you have a tabbed view in some deeper part of your application? The tabbed content is a Server Component that does some expensive things and renders the correct content. Now you might want to invalidate only that deeply nested SC, but not the ones higher up in the tree, so any setup or prefetching you normally do when the entire page is rendered, might not happen.

I think the above is one reasoning around not supporting context in RSCs btw, it's a different model. I'm not sure yet what patterns will emerge etc, but I think from a library/API design point of view it makes sense to avoid abstractions that are shared between different Server Components.

**Reply by amannn** [reply]
Thank you for your thoughts, these are really helpful!

> We set up some Providers at the top and trust they are always there, providing their context. With Server Components, that might not always be the case (in the future?)

This reminds me of the initial RSC demo, where a client side state was used to update a RSC. AFAIK this pattern is currently not possible in Next.js, but likely something to anticipate, based on that demo.

You might have a point with the statement that a different API might make more sense for RSC. I'm still a bit in a frame where I like to think of a unified "React environment", but e.g. with async/await and `use` in components (not to speak of `'use client';`), it might be helpful to be really conscious about what happens on the server and what on the client.

**Reply by TkDodo** [reply]
> In RSCs, because fetch is deduplicated automatically, data will be in sync between RSCs anyway

this assumes that all we do with TanStack query is using "fetch", which is not necessarily how I see it. There might be other async operations that will not get deduped automatically.

> What if you have a tabbed view in some deeper part of your application? The tabbed content is a Server Component that does some expensive things and renders the correct content. Now you might want to invalidate only that deeply nested SC, but not the ones higher up in the tree, so any setup or prefetching you normally do when the entire page is rendered, might not happen.

I don't really see how this scenario wouldn't work with a `queryClient`. Something like:

```
const MyTabServerComponent = () => {
  const queryClient = new QueryClient()
  const data = await queryClient.fetchQuery(...)

  return (
    <Suspense fallback="Loading...">
      <HydrationBoundary state={dehydrate(data)}>
        <SomeMoreJsxWithUseQueryInside />
      </HydrationBoundary>
    </Suspense>
  );
}
```

**Reply by Ephem** [reply]
> this assumes that all we do with TanStack query is using "fetch", which is not necessarily how I see it

That's a very good point! We might probably want to have a request scoped `queryClient` after all then, but maybe with a slightly different API for creating/accessing it?

> I don't really see how this scenario wouldn't work with a queryClient

It would! This would also work to share the `queryClient` between RSCs (as can be seen in [this codesandbox by @efilion](https://codesandbox.io/p/github/efilion/next13-react-query/hydrated-content?file=%2Fapp%2FgetQueryClient.tsx&selection=%5B%7B%22endColumn%22%3A62%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A1%7D%5D)):

```jsx
import { QueryClient } from "@tanstack/query-core";
import { cache } from "react";

export const getQueryClient = cache(() => new QueryClient());
```

But this wouldn't necessarily:

```jsx
async function TopServerComponent() {
  const article = await getQueryClient().prefetchQuery('article');

  // ...
}

// Other file
async function DeeperServerComponent() {
  const article = getQueryClient().getQueryData('article');

  // ...
}
```

Not that this is a common pattern, but in the traditional model it would kind of work because `TopServerComponent` would always run first, but in this model, `DeeperServerComponent` might render on it's own if only that gets invalidated.

The main point I was trying to make though was more about why context wouldn't work with RSCs. :smiley:

**Reply by Ephem** [reply]
Actually, a better example is probably this:

```jsx
async function TopServerComponent() {
  const queryClient = getQueryClient(INITIAL_OPTIONS);

  // ...
}

// Other file
async function DeeperServerComponent() {
  // Easy to believe/think I have already initiated this with INITIAL_OPTIONS
  const queryClient = getQueryClient();

  // ...
}
```

**Reply by TkDodo** [reply]
> That's a very good point! We might probably want to have a request scoped queryClient after all then, but maybe with a slightly different API for creating/accessing it?

I think this is where "server context" would come into play? I don't think it exists yet though 😅 . But yes, something like this would be nice.

> export const getQueryClient = cache(() => new QueryClient());

oh that's neat. a request-scoped server context would still probably be the best solution to initializing a queryClient only once.

**Reply by amannn** [reply]
FYI, `createServerContext` is available in the latest version of React (although the TypeScript types are not there yet):

```tsx
import {createServerContext} from 'react';

const Context = createServerContext(globalName: string, value: any);
```

The value is transmitted for every request between the server and client, so it needs to be really small and also the value needs to be serializable. So I think putting a `queryClient` there might not work. From my current understanding, using `cache()` to create a request-scoped `QueryClient` might be the better choice.

For reference, I'm [working on the same problem currently for `next-intl`](https://github.com/amannn/next-intl/pull/149), an i18n library of Next.js I'm maintaining.

**Reply by Ephem** [reply]
Yeah, `createServerContext` is interesting, I'm not quite sure how it fits into the big picture yet. Maybe one interesting avenue to explore would be to find a minimal format for keeping track of which queries the client already has fresh data for, so the server can avoid prefetching and de/rehydrating those when not necessary?

**Reply by TkDodo** [reply]
I think one case would be to provide a `queryClient` so that you can prefetch wherever you want?

**Reply by cyco130** [reply]
I think that's exactly what you **can't** do. The value has to be serializable.

**Reply by TkDodo** [reply]
Oh 🤦‍♂️ I totally misread that. I thought the context would be request scoped

**Comment by cyco130** [comment]
Hi,

I'm the author of [Rakkas](https://rakkasjs.org), a Vite-based React metaframework with some unique cutting edge features. It doesn't support RSC yet but it has supported streaming SSR with Suspense for quite sometime now (before Next.js did).

I have a [Rakkas-Tanstack Query integration example repo](https://github.com/rakkasjs/tanstack-query) in which I devised a scheme to make Tanstack Query work with streaming SSR. I thought it might be of interest in this discussion. I just added some documentation on how it works in the readme but, basically, instead of relying on Tanstack Query's own deydration/hydration features, I handle the serialization myself: Every time a query is resolved on the server, I inject a script into React's HTML stream that calls `queryClient.setQueryData` with the serialized query data. I haven't done extensive testing but it seems to work well. The implementation uses some Rakkas specific features but the `emitBeforeSsrChunk` can be implemented in any framework that supports streaming SSR by wrapping the SSR stream and injecting something before each chunk.

It may also be of interest that Rakkas has its own implementation of [`useQuery`](https://rakkasjs.org/guide/use-query) and some other Tanstack Query APIs. This implementation is Suspense-only and supports only a subset of the features. It's used under the hood for Rakkas's unique inline RPC mechanism [`useServerSideQuery`](https://stackblitz.com/edit/use-ssq?file=src%2Froutes%2Findex.page.tsx). Rakkas's implementation doesn't have the concept of hydration/dehydration and it uses a system similar to the above.

Lastly, you may want to check out @brillout's [`react-streaming`](https://github.com/brillout/react-streaming), a library to simplify the setup for implementing React streaming SSR. It comes with a `useAsync` hook which is similar to `useQuery` in many ways. It uses a mechanism very similar to the one I used to send server-side query data to the client.

**Reply by Ephem** [reply]
Thank you for this, I'm familiar with `react-streaming`, but hadn't heard of Rakkas before, I'll definitely check it out!

I think optimally, React would provide a standardized way to pass "data discovered during rendering" from server->client, similar to how I know there has been talk about having built-in hooks for early preloading of images etc discovered during serverside rendering for example. That would move this concern from the metaframework-level to the library level and would let us create a public API for RQ that just works wherever you run it, without setup.

If that doesn't happen (or before it does, or maybe even if it does), what we might do is also to provide an option where you can pass in your own `emitBeforeSsrChunk` so to speak and as long as that matches some interface we decide, we use that to automatically de/rehydrate data discovered during rendering to the client. This would be an opt in feature for frameworks that can support it, but required for Streaming Suspense SSR.

**Reply by Ephem** [reply]
Just to play devils advocate for a bit, there are arguments against even supporting this at all though.

If SSR hydration would support Promise de/rehydration just as planned for RSCs, we wouldn't need this "type" of streaming hydration as long as all queries start prefetching before rendering starts (doesn't matter if it's via RSCs or "just" some router level thing). Why is this an argument against hydrating data discovered during rendering? Well, supporting automatic de/rehydration of data discovered during rendering _will_ lead to request waterfalls. When on the server, this has the same constraints as in RSCs really and doesn't matter (as much), but since Client Components also run in the client with higher latency, these waterfalls will hurt (just as is common today).

Personally I think this is way too prescriptive of a stance to take though, what we should focus on instead is finding ways to help developers notice and fix these waterfalls.

**Reply by cyco130** [reply]
I'm afraid "a standardized way to pass data from server to client" may never materialize. Historically the React team has kept their distance with these matters and left them to metaframeworks. On the other hand RSC _requires_ a lot of machinery beyond React's "not a framework, just a library" model so who knows.

Before streaming SSR, we were collecting all the data and inserting a single script into the HTML. Injecting scripts into the HTML stream as new data becomes available is the logical extension of that.

BTW do you see any flaw in this approach for using Tanstack Query with streaming SSR? I'm gonna do more complete tests in the future but is there any blockers or downsides that I'm not aware of (e.g. related to Tanstack Query's internals) or is this a good avenue until something official becomes available?

**Reply by cyco130** [reply]
Re waterfalls: I see how it can be a real problem in some apps. There's a trade-off between route level data fetching (no waterfalls) and component level data fetching (ease of use and composability). I think it should be left to the developer which point in the spectrum they want to be.

Of course the library can provide some guidance though: To avoid waterfalls, one would start fetching a query without awaiting the promise at the route level (e.g. `getServerSideProps` and similar) when it's known that it will be used in that route to parallelize things. And `useQuery` can optionally throw a warning when an SSR query wasn't already fetching at the time it was called. Also a query that started but was never used could also cause a warning for example.

**Reply by Ephem** [reply]
> Historically the React team has kept their distance with these matters and left them to metaframeworks

I think this has been a good stance historically, but on the other hand I see them forming a lot more opinions on these matters when it comes to these new features. Or another way to say it, I think these new features paves the way for how they have always envisioned doing these things. I think they want to get it juust right before including things in the framework itself and I have a lot of respect for that taking time. :smiley: We'll see! Will be interesting to read the cache RFC when that hits.

> do you see any flaw in this approach for using Tanstack Query with streaming SSR?

Without having had time to look properly, not that I know off. I mean, hydration currently isn't perfect, we always do it in render even in cases that can trigger observable side effects and doesn't use the `getServerSnapshot` properly so hydrating more often in more places might more easily surface edge cases and current bugs, but even if so they shouldn't too bad even now.

(As an aside, here's an old discussion and demo by @artem-malko of this appropach that also has some context and comments, a year old so might be slightly outdated by now: https://github.com/TanStack/query/discussions/2942)

**Reply by artem-malko** [reply]
> so might be slightly outdated by now

No @Ephem, it works the same way, there are no any updates) Actually, I'm waiting for the Cache API from React to make a new demo. But there is no any updates about this API( Or it's just somewhere in private chats, I don't know.

By the way @cyco130, we're thinking in the same way) The only one important difference — I tried to use react-query without server/client specific for a developer. And in my dreams I use `useQuery` just everywhere I want in the future, without thinking about server/client context.

**Reply by Ephem** [reply]
Nice! I hadn't read through it myself in a while so I wasn't sure. :smiley:

**Reply by brillout** [reply]
> I think optimally, React would provide a standardized way to pass "data discovered during rendering" from server->client, similar to how I know there has been talk about having built-in hooks for early preloading of images etc discovered during serverside rendering for example. That would move this concern from the metaframework-level to the library level and would let us create a public API for RQ that just works wherever you run it, without setup.

Exactly. And this is actually happening, see [React RFC - injectToStream](https://github.com/reactjs/rfcs/pull/219), most notably sebmarkbage's comment `Therefore our focus has been on adding higher level replacements for each use case.`.

> finding ways to help developers notice and fix these waterfalls.

I think it's possible. A cheap but fairly reliable way to detect waterfalls is to detect requests that begin after another request has finished. There can be false positives, e.g. if something delays the rendering of a data-fetching component so much that it starts rendering (and therefore its data-fetching) after another data request has finished. I think a fairly reliable workaround would be this: consider a request to be a waterfall if it starts after another request finished AND if it starts 100 milliseconds after rendering of the page started. 

Ideally, there would be a 100% reliable way to detect waterfalls, but I can't think of a way that don't invovle some deep integration with React.

**Reply by Ephem** [reply]
> Exactly. And this is actually happening, see https://github.com/reactjs/rfcs/pull/219, most notably sebmarkbage's comment Therefore our focus has been on adding higher level replacements for each use case..

Thanks for reminding me about this one, I guess the big questions are0 not _if_, but when they will get around to the data fetching usecase and what it will look like.

> Ideally, there would be a 100% reliable way to detect waterfalls, but I can't think of a way that don't invovle some deep integration with React.

This would be nice! I think heuristics might do fine here though. Something slightly different that would be easier to implement would be to help find _queries that were not preloaded_. This covers less usecases, but might also be helpful.

**Reply by brillout** [reply]
> when they will get around to the data fetching usecase

If the new [`use()` hook](https://github.com/reactjs/rfcs/pull/229) allows to the user to determine the cache lifecycle then it'll do the trick. ~~Although I've doubts that such option will be available.~~

**Reply by brillout** [reply]
Actually, it will support cache control: https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md#caveat-data-requests-must-be-cached-between-replays.

Neat, so I think `use()` will completely replace `react-streaming`. Neat neat.

**Reply by artem-malko** [reply]
Hm, @brillout could you explain, how `use()` can replace `react-streaming`?

**Reply by Ephem** [reply]
I think `use()` only deals with consuming for example a fetch that was started on the server, since it can take a Promise and read if resolved or suspend if not. It doesn't really deal with how to pass that Promise from the server to the client though, but that is mentioned here: [https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md#passing-a-promise-from-a-server-component-to-a-client-component](https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md#passing-a-promise-from-a-server-component-to-a-client-component)

That's just for passing a Promise from a Server Component though, the thing I'm still unclear on is passing directly from the "normal" SSR pass to the client, for any query that was not prefetched in an SC, but instead discovered during the rendering pass.

**Comment by cadoopmin029** [comment]
To whom this may concern,

I do not know why I am receiving all of these notifications. I did not
subscribe to follow or receive anything. Can someone advise as to how this
may have happened?

Thanks

**Reply by Ephem** [reply]
Not sure why you are getting notifications, but it usually says why right next to the Unsubscribe-button that should also fix the problem for you:

![image](https://user-images.githubusercontent.com/7420167/214260461-4ffe662a-3c94-4f3d-8a9e-8141db39265f.png)

**Comment by tonyxiao** [comment]
Prefetching now can happen in multiple server components, how would you collect all of them together to the right `HydrationBoundary` just once? Or is there a way for `HydrationBoundary` to be able to merged internally such that it can be called multiple times?

**Comment by kbariotis** [comment]
@Ephem thank you for this explanation and the work you've put into it. There's one missing piece that I'm struggling to understand and maybe the implementation has changed since this comment was written. If promises are being passed down to components from the queryClient and `use` is being used to consume those promises, why do we need to the `prefetch`? Shouldn't that be handled automatically?

It also confuses me in [the docs](https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr#experimental-streaming-without-prefetching-in-nextjs) about the elimination of prefetching which will have the waterfall as an effect but I don't see how all these tie up together.

Thank you

**Reply by Ephem** [reply]
Hi! This is a common question and it is a bit tricky to wrap ones head around.

Prefetch is needed to start the promise as early as possible. By doing this in a server component, any prefetches in nested server components are guaranteed to either run in parallel, or at least only waterfall _on the server_, because they are guaranteed to run there.

With the experimental approach _without_ explicit prefetch, the fetch is started in a _client_ component, when React renders `useSuspenseQuery`. On the first page load, this happens on the server, because client components are server rendered on page load. Client components are _not_ rendered on the server at all on subsequent navigations however, and now if you have `useSuspenseQuery` in two components that are nested, you now have a clientside waterfall which hurts performance.

This might seem like not a huge deal and of course the tradeoffs depend on your specific situation, but in practice I've found many apps have more waterfalls than they think with the prefetch-less approach, especially when you also take into account that code has to load before the `useSuspenseQuery` can run.

Did that make things any clearer, or just more confusing? 😄
